### PR TITLE
perf: faster component serialization/deserialization

### PR DESCRIPTION
**Pull request by Lymdun**
NetworkIdentity DirtyComponentsMask code removed entirely. OnSerialize now includes the component index as byte before serializing each component. Faster because we avoid GetDirtyComponentsMask() and GetSyncModeObserversMask() calculations. Increases allowed NetworkBehaviour components from 64 to 255. Code is significantly more simple. (#2331)

* perf: NetworkIdentity DirtyComponentsMask code removed entirely. OnSerialize now includes the component index as byte before serializing each component. Faster because we avoid GetDirtyComponentsMask() and GetSyncModeObserversMask() calculations. Increases allowed NetworkBehaviour components from 64 to 255. Bandwidth is now smaller if <8 components, and larger if >8 components. Code is significantly more simple.

* Update Assets/Mirror/Runtime/NetworkIdentity.cs

Co-authored-by: James Frowen <jamesfrowendev@gmail.com>

Co-authored-by: James Frowen <jamesfrowendev@gmail.com>

### CODE REVIEW

File: Assets/Mirror/Runtime/NetworkIdentity.cs [hunk 1]
```diff
@@ -817,45 +817,37 @@ void OnSerializeSafely(NetworkBehaviour comp, NetworkWriter writer, bool initial
...
                 {
                     if (logger.LogEnabled()) logger.Log("OnSerializeAllSafely: " + name + " -> " + comp.GetType() + " initial=" + initialState);
 
+                    // remember start position in case we need to copy it into
+                    // observers writer too
+                    int startPosition = ownerWriter.Position;
+
+                    // write index as byte [0..255]
+                    ownerWriter.WriteByte((byte)i);
+
```
Comment (paulpach) [inline Assets/Mirror/Runtime/NetworkIdentity.cs:848]:
```suggestion
```

Remove this,  we have the barrier system instead

File: Assets/Mirror/Runtime/NetworkServer.cs [hunk 2]
```diff
@@ -393,31 +393,17 @@ public void SendToAll<T>(T msg, int channelId = Channel.Reliable)
...
             foreach (NetworkIdentity identity in DirtyObjects)
             {
                 if (identity != null)
-                {
-                    identity.ServerUpdate();
-
-                    if (identity.StillDirty())
-                        DirtyObjectsTmp.Add(identity);
-                }                
+                    identity.ServerUpdate();
```
Comment (paulpach) [inline Assets/Mirror/Runtime/NetworkServer.cs:405]: Put this function back.

This has nothing to do with the change.

This was an optimization we had that prevents us from looping all the objects looking for dirty ones,  instead we only serialize the ones that are dirty.

So for us,  this is a regression

File: Assets/Mirror/Runtime/NetworkIdentity.cs [hunk 3]
```diff
@@ -817,45 +817,37 @@ void OnSerializeSafely(NetworkBehaviour comp, NetworkWriter writer, bool initial
...
                 // -> always serialize if initialState so all components are included in spawn packet
                 // -> note: IsDirty() is false if the component isn't dirty or sendInterval isn't elapsed yet
+                NetworkBehaviour comp = components[i];
                 if (initialState || comp.IsDirty())
                 {
                     if (logger.LogEnabled()) logger.Log("OnSerializeAllSafely: " + name + " -> " + comp.GetType() + " initial=" + initialState);
 
+                    // remember start position in case we need to copy it into
+                    // observers writer too
+                    int startPosition = ownerWriter.Position;
```
Comment (paulpach) [inline Assets/Mirror/Runtime/NetworkIdentity.cs:844]:
```suggestion
```

I don't think we use this at all.  Our barrier system does not need that

File: Assets/Mirror/Runtime/NetworkIdentity.cs [hunk 4]
```diff
@@ -881,66 +873,6 @@ void OnSerializeSafely(NetworkBehaviour comp, NetworkWriter writer, bool initial
...
-        /// <returns></returns>
-        internal bool StillDirty()
-        {
-            foreach (NetworkBehaviour behaviour in NetworkBehaviours)
-            {
-                if (behaviour.StillDirty())
-                    return true;
-            }
-            return false;
-        }
```
Comment (paulpach) [inline Assets/Mirror/Runtime/NetworkIdentity.cs:914]: this is used in the optimization you reverted below

File: Assets/Mirror/Runtime/NetworkServer.cs (hunk 2 above)
Comment (Lymdun) [reply Assets/Mirror/Runtime/NetworkServer.cs]: fixed

File: Assets/Mirror/Runtime/NetworkServer.cs [hunk 5]
```diff
@@ -411,13 +411,13 @@ internal void Update()
 
                     if (identity.StillDirty())
                         DirtyObjectsTmp.Add(identity);
-                }                
-            }
```
Comment (paulpach) [inline Assets/Mirror/Runtime/NetworkServer.cs:415]: you deleted this

**Pull request merged**